### PR TITLE
[Ready]Removes martial arts book granters and adds a more futuristic way to learn them

### PR DESCRIPTION
**Pull request by wesoda25**
## About The Pull Request

Adds a rough framework for a new way to learn things, but for now its sole purpose is for martial arts. 

Adds the AED (Automatic Education Device) and education chips. Simply insert an education chip into the AED and activate it to instantly learn the subject. 

## Why It's Good For The Game

Reading a scroll and learning a martial art in 5 seconds is pretty dumb. Its more futuristic now and more at home in the game

## Changelog
:cl:
add: Automatic Education Device (AED) and Education Chips, imply insert an education chip into the AED and activate it to instantly learn the subject. For now you can only learn martial arts (CQC, Krav Maga, Sleeping Carp, Plasma Fist - Mush not included). You can find one in the curators locker, but its no good without a chip. 
del: Martial Arts Book Granters, Krav Maga Implants
imageadd: Automatic Education Device and Education Chip sprites
/:cl:

Oh and its not in violation of antag freeze orange man said its fine

### CODE REVIEW

**Comment by wesoda25** [comment]
my cl is shit ill fix that

**Comment by carlarctg** [comment]
this is dumber than scrolls because scrolls at least let you suspend your disbelief

also ruins carp lore

**Comment by wesoda25** [comment]
wtf is carp lore

**Comment by carlarctg** [comment]
> wtf is carp lore

carp lore

**Comment by PKPenguin321** [comment]
im pretty sure we already have implanters that teach you the martial art in the code

**Comment by wesoda25** [comment]
Yeah I saw that when I was basically done but that thing is dumb too idk if i should consolidate into this

course maybe mine is the dumb one but whatever

**Comment by Shadowflame909** [comment]
Stand discs already exist in jojo wesoda!1!

**Comment by TheVekter** [comment]
Is it weird that my first thought was I KNOW KUNG FU WOAH as opposed to STAND DISCS?

Whatever, either way, this is cool as hell.

**Comment by tmtmtl30** [comment]
classic sci-fi stuff, i love it -- kind of surprising we didn't have skill neurochips already

**Comment by PKPenguin321** [comment]
>Is it weird that my first thought was I KNOW KUNG FU WOAH as opposed to STAND DISCS?

It's the opposite actually

**Comment by Tlaltecuhtli** [comment]
it should be a wikihow tutorial you can watch on the ipad

**Comment by Tharcoonvagh** [comment]
This will make people using the AEDs valid for the metagamers, ples fix

**Comment by Fikou** [comment]
>be carpman
>learn legendary technique, pass it onto a scroll so that it may be passed onto generations by the syndicate
>get replaced by a technology instant dabber

**Comment by wesoda25** [comment]
I removed krava maga implants and put them into here, among other things. I maintain that Sleeping Carp scroll is stupid and has no established lore (not that it matters anyways), however if enough people want it to remain in scroll format I suppose I could do that. 

Still think my names are dumb and am open to suggestions in that department.

**Comment by Shadowflame909** [comment]
Only thing is, if they're "Chips" cant someone just do surgery on your head and  recieve the martial arts from inside of the chip?

Unless the lore is "Dissolving Chip."

Tldr; Is it becoming something like a removable implant? Or is it still staying within the mind datum.

**Comment by wesoda25** [comment]
Its all in the mind.

You put on the AED, put  a chip into the AED, it "transports" the information from the chip to your head, leaving the chip empty. The code itself has not changed that much, just the medium.

**Comment by AnonymousNow** [comment]
We have mysticism and magic in the setting, and the natures of our martial arts are supernatural. I don't think that this is a needed change.

**Comment by wesoda25** [comment]
eh see I personally dislike that stuff, I've always thought its pretty stupid (and that wizard is one of our most lacking). But if I'm alone in that outlook I suppose I'll just close.

File: code/game/objects/items/devices/education_chips.dm [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+/obj/item/education_chip/cqc
+  name = "Close-Quarters-Combat Education Chip"
+  desc = "An education chip containing the martial art of CQC. Insert into an Automatic Education Device to learn."
+  subject = /datum/martial_art/cqc
+  subject_name = "Close-Quarters-Combat"
+  greet = "<span class='boldannounce'>You've mastered the basics of CQC.</span>"
+
+/obj/item/education_chip/cqc/Initialize()
+  . = ..()
+  add_overlay("edchip_cqc")
```
Comment (tralezab) [inline code/game/objects/items/devices/education_chips.dm]: if you're gonna do this for each one just add a variable

File: code/game/objects/items/devices/education_chips.dm [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+  for(var/obj/item/education_chip/E in src.contents)
+    if(!E.subject)
+      to_chat(usr, "<span class='warning'>The Education Chip doesn't have a subject!</span>")
+      return
+    var/datum/martial_art/MA = new E.subject
+    if(usr.mind.has_martialart(initial(MA.id)))
+      to_chat(usr,"<span class='warning'>You already know [E.subject_name]!</span>")
+      return
+    to_chat(usr, "<span class='notice'>Your head whirls with a sudden rush of information as you instantly learn about [E.subject_name]!</span>")
+    to_chat(usr, "[E.greet]")
```
Comment (tralezab) [inline code/game/objects/items/devices/education_chips.dm]: there is no check to see if there is actually a greet. any number of these suggestions are good, but at least one should be done so it doesn't send an empty message to chat:

- add a default greet
- add a check to not send this if there is no greet
- give krav maga a damn greet

File: code/game/objects/items/devices/education_chips.dm [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+/obj/item/education_chip/plasma_fist/Initialize()
+  . = ..()
+  add_overlay("edchip_pfist")
+
+/obj/item/education_chip/krav_maga
+  name = "Krava Maga Education Chip"
+  desc = "An education chip containing the martial art of Krav Maga. Insert into an Automatic Education Device to learn."
+  subject = /datum/martial_art/krav_maga
+  subject_name = "Krav Maga"
+
```
Comment (tralezab) [inline code/game/objects/items/devices/education_chips.dm]: right now this sends an empty message to chat because no greet

File: code/game/objects/items/devices/education_chips.dm [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+/obj/item/clothing/head/AED/verb/educate()
+  set src in usr
+  if(!src.contents.len)
+    to_chat(usr, "<span class='warning'>There is nothing loaded in the [src]!</span>")
+    return
+  for(var/obj/item/education_chip/E in src.contents)
+    if(!E.subject)
+      to_chat(usr, "<span class='warning'>The Education Chip doesn't have a subject!</span>")
+      return
```
Comment (tralezab) [inline code/game/objects/items/devices/education_chips.dm]: if you're gonna make this a for loop these should be continues. pockets can hold two chips, randomly ending sometimes depending on the order you added the chips is bad.

File: code/game/objects/items/devices/education_chips.dm [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+    var/datum/martial_art/MA = new E.subject
+    if(usr.mind.has_martialart(initial(MA.id)))
+      to_chat(usr,"<span class='warning'>You already know [E.subject_name]!</span>")
+      return
+    to_chat(usr, "<span class='notice'>Your head whirls with a sudden rush of information as you instantly learn about [E.subject_name]!</span>")
+    to_chat(usr, "[E.greet]")
+    MA.teach(usr)
+    usr.log_message("learned the subject [E.subject_name] ([MA])", LOG_ATTACK, color="orange")
+    qdel(E)
+    new /obj/item/education_chip(src)
```
Comment (tralezab) [inline code/game/objects/items/devices/education_chips.dm]: the old books had a var for multiple uses for admins, i suggest you do the same

File: code/game/objects/items/devices/education_chips.dm [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+
+
+////Education Chips////
+
+/obj/item/education_chip
+  name = "Empty Education Chip"
+  desc = "An empty education chip."
```
Comment (tralezab) [inline code/game/objects/items/devices/education_chips.dm]: chances are traitors will trash or otherwise discard these when used, new players should probably know what these do beyond "hey look an empty chip??"

**Review by tralezab — COMMENTED** [review]
just some things i noticed.

**Comment by tralezab** [comment]
personally i'm gonna miss the cqc book, it was much cooler that the cqc book beeped and exploded when used again. still, every other change is good. we should definitely be quarantining magic to wizard only, it really breaks the theme.

**Comment by 81Denton** [comment]
Not a huge fan of this since martial arts scrolls and exploding CQC books have more flavor than just slapping a chip into a device and using it.

**Comment by carlarctg** [comment]
> personally i'm gonna miss the cqc book, it was much cooler that the cqc book beeped and exploded when used again. still, every other change is good. we should definitely be quarantining magic to wizard only, it really breaks the theme.

cqc book was never magic tho

File: code/game/objects/items/devices/education_chips.dm (hunk 4 above)
Comment (wesoda25) [reply code/game/objects/items/devices/education_chips.dm]: It can only hold one chip

**Comment by wesoda25** [comment]
Course bugs aren't a valid reason for removal, but did cqc book ever actually explode? I I think it did for some period of time but hasnt worked recently, or when I started playing.

File: code/game/objects/items/devices/education_chips.dm (hunk 6 above)
Comment (wesoda25) [reply code/game/objects/items/devices/education_chips.dm]: Put a little more info there, but I don't want to add "suspicious-looking" or anything blatant. Possibly one day could be utilized for teaching skills or something idk.

**Comment by SkeletalElite** [comment]
> Course bugs aren't a valid reason for removal, but did cqc book ever actually explode? I I think it did for some period of time but hasnt worked recently, or when I started playing.

Yeah it hasn't been exploding for a while now.

File: code/game/objects/items/devices/education_chips.dm (hunk 6 above)
Comment (tralezab) [reply code/game/objects/items/devices/education_chips.dm]: makes sense to me, i like it

**Comment by tralezab** [comment]
the cqc book works like spellbooks now, where if you use it, it beeps ominously and the next person to use it gets exploded

**Comment by optimumtact** [comment]
I like this, feels way more matrix and scifi than scrolls

File: code/game/objects/items/devices/education_chips.dm [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+  pocket_storage_component_path = /datum/component/storage/concrete/pockets/tiny/AED
+
+/obj/item/clothing/head/AED/item_action_slot_check(slot)
+  if(slot == SLOT_HEAD)
+    return 1
+
+/obj/item/clothing/head/AED/ui_action_click(mob/user, action)
+  educate()
+
+/obj/item/clothing/head/AED/verb/educate()
```
Comment (AnturK) [inline code/game/objects/items/devices/education_chips.dm]: Add user parameter and use it instead of usr.

File: code/game/objects/items/devices/education_chips.dm [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+  w_class = WEIGHT_CLASS_TINY
+  var/subject
+  var/subject_name
+  var/greet
+  var/unlimited = FALSE //for admins if they want unlimited uses
+  var/overlay
+
+/obj/item/education_chip/Initialize()
+  . = ..()
+  add_overlay(overlay)
```
Comment (AnturK) [inline code/game/objects/items/devices/education_chips.dm]: Either check if this is null or add a default. Also should probably be in update_overlays

**Review by AnturK — CHANGES_REQUESTED** [review]

**Comment by AnturK** [comment]
I feel it would be more interesting if the teachers were available somewhere publicly (library?) for other skills and uplinks only provided illegal material.

**Comment by davethwave** [comment]
We could just keep both and use one or the other depending on theme.

**Comment by optimumtact** [comment]
compromise PAH

File: code/game/objects/items/devices/education_chips.dm (hunk 7 above)
Comment (wesoda25) [reply code/game/objects/items/devices/education_chips.dm]: idk what that is tbqh

Comment (wesoda25) [reply code/game/objects/items/devices/education_chips.dm]: the thing inside parentheses? 
so 
`/obj/item/clothing/head/AED/verb/educate(mob/user)`?

File: code/game/objects/items/devices/education_chips.dm (hunk 8 above)
Comment (wesoda25) [reply code/game/objects/items/devices/education_chips.dm]: ok ill do that when i do the other

**Comment by wesoda25** [comment]
> I feel it would be more interesting if the teachers were available somewhere public (library?) for other skills and uplinks only provided illegal material.

Might consider putting one in the library, maybe a stationary chair type thing you strap yourself into. But I think traitors should still get one otherwise valid hunters will just camp it.

**Comment by optimumtact** [comment]
if you don't make the library one look like the matrix chair I will be very disappointed

File: code/game/objects/items/devices/education_chips.dm (hunk 7 above)
Comment (AnturK) [reply code/game/objects/items/devices/education_chips.dm]: Yes, usr should only be used in top level Topic calls if possible.

Comment (wesoda25) [reply code/game/objects/items/devices/education_chips.dm]: How would I do `set src in usr` then? It only allows for usr.

Comment (AnturK) [reply code/game/objects/items/devices/education_chips.dm]: I forgot to mention this shouldn't be a verb and that part is unnecessary.

**Comment by wesoda25** [comment]
Sorry to disappoint orange man, but A: im lazy and didn't wanna do that and B: don't want a carbon copy of matrix shit anyways. 

Now I tentatively say this is actually ready barring something else comes up. (still taking name suggestions!!)

File: code/game/objects/items/devices/education_chips.dm (hunk 8 above)
Comment (wesoda25) [reply code/game/objects/items/devices/education_chips.dm]: couldnt figure out update_overlays so i just made a check, not perfect I guess but better

File: code/game/objects/items/devices/education_chips.dm (hunk 7 above)
Comment (wesoda25) [reply code/game/objects/items/devices/education_chips.dm]: think i got what you wanted

File: code/game/objects/items/devices/education_chips.dm [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+  w_class = WEIGHT_CLASS_TINY
+  var/subject
+  var/subject_name
+  var/greet
+  var/unlimited = FALSE //for admins if they want unlimited uses
+  var/overlay
+
+/obj/item/education_chip/Initialize()
+  . = ..()
+  if(overlay)
```
Comment (AnturK) [inline code/game/objects/items/devices/education_chips.dm]:
```suggestion
  update_icon()

/obj/item/education_chip/update_overlays()
  . = ..()
  if(overlay)
    . += overlay
```

Comment (AnturK) [reply code/game/objects/items/devices/education_chips.dm]: What i meant by update_overlays.

**Comment by wesoda25** [comment]
idk anturk that stuff dont work

**Comment by wesoda25** [comment]
ohhh nooooooo

**Comment by wesoda25** [comment]
ooooooooohhhhhhhhhhh noooooooooooooooooooooooooooooooooooooooooooooooo

**Comment by wesoda25** [comment]
this is so fucked i dont even know whats fucked anymore

**Comment by wesoda25** [comment]
yeah idk man thats the best you're getting from me

**Comment by PKPenguin321** [comment]
You unchecked the files during conflict resolution you fool

**Comment by wesoda25** [comment]
pk i literally have no idea what that even means but yeah i agree with the second part

**Comment by wesoda25** [comment]
yknow i think ill just make another pr

**Comment by PKPenguin321** [comment]
when you make a conflict-fixing commit it will say that something like "10,000 files changed" and newbies will look at that and say "wtf i didnt change 10,000 files" so they will exclude them from the commit, and this is a TRAP that will break your branch, because to fix conflicts it needs those files listed as changed
